### PR TITLE
[2201.5.0-stage] Fix completions inside capturing group

### LINKED ISSUE #40099: [Bug]: RegExp quantifier completion item appends unintended characters
### Description

https://user-images.githubusercontent.com/15815199/229745443-9b3864e9-3add-454f-9021-008dff131ecc.mov

### Steps to Reproduce

_No response_

### Affected Version(s)

2201.4.1

### OS, DB, other environment details and versions

_No response_

### Related area

-> Editor

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #40128: [Bug]: Regex suggest invalid completion
### Description

```
import ballerina/lang.regexp;
public function main() {
    regexp:RegExp _ = re ``<cursor>;
}
```
Gives invalid completions as `()`,` []`, `^`

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes #40099
Fixes #40128

Related PRs: #40103

https://user-images.githubusercontent.com/46857198/229776184-81dc7ca5-8c70-4aff-bf45-7191414947d6.mov

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java [hunk 1]
```diff
@@ -481,6 +489,12 @@ private static boolean isReAtom(NonTerminalNode nodeAtCursor, BallerinaCompletio
                 || nodeAtCursor.kind() == SyntaxKind.RE_CAPTURING_GROUP;
     }
 
+    private static boolean isBeginOfReCapturingGroup(NonTerminalNode nodeAtCursor, BallerinaCompletionContext ctx) {
```
Comment (pcnfernando) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java:492]:
```suggestion
    private static boolean isCapturingGroupStart(NonTerminalNode nodeAtCursor, BallerinaCompletionContext ctx) {
```

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java [hunk 2]
```diff
@@ -717,7 +731,7 @@ public SnippetBlock get() {
     /**
      * Visitor to find the template expression node.
      */
-     static class RegexTemplateNodeFinder extends NodeVisitor {
+    static class RegexTemplateNodeFinder extends NodeVisitor {
```
Comment (pcnfernando) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java:734]: Don't format the whole file. Format just the parts you are updating

File: language-server/modules/langserver-core/src/test/resources/completion/template_expression/source/regex_template_expression_source13.bal [hunk 3]
```diff
@@ -0,0 +1,5 @@
+function testRegexp() {
+    re `()`;
+    
+    re `(\(\))`;
+}
```
Comment (pcnfernando) [inline language-server/modules/langserver-core/src/test/resources/completion/template_expression/source/regex_template_expression_source13.bal:5]: Missing eof

File: language-server/modules/langserver-core/src/test/resources/completion/template_expression/source/regex_template_expression_source13.bal [hunk 4]
```diff
@@ -0,0 +1,5 @@
+function testRegexp() {
+    re `()`;
+    
+    re `(\(\))`;
```
Comment (pcnfernando) [inline language-server/modules/langserver-core/src/test/resources/completion/template_expression/source/regex_template_expression_source13.bal:4]: Add a possible quantifier case within the group

**Review by pcnfernando — APPROVED** [review]

**Review by sanjana — COMMENTED** [review]
Minor formatting suggestion. Other than that everything else LGTM.

File: language-server/modules/langserver-core/src/test/resources/completion/template_expression/source/regex_template_expression_source13.bal [hunk 5]
```diff
@@ -0,0 +1,7 @@
+function testRegexp() {
+    re `()`;
+
```
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/test/resources/completion/template_expression/source/regex_template_expression_source13.bal:3]: Did like this to improve the readability.

**Review by sanjana — APPROVED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java [hunk 6]
```diff
@@ -481,6 +490,12 @@ private static boolean isReAtom(NonTerminalNode nodeAtCursor, BallerinaCompletio
                 || nodeAtCursor.kind() == SyntaxKind.RE_CAPTURING_GROUP;
     }
 
+    private static boolean isCapturingGroupStart(NonTerminalNode nodeAtCursor, BallerinaCompletionContext ctx) {
+        return nodeAtCursor.kind() == SyntaxKind.RE_CAPTURING_GROUP
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java:494]: Is this conidtion valid for `re (  <cursor>)` ?

Comment (LakshanWeerasinghe) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/RegexpCompletionProvider.java:494]: No. It should not valid.

**Review by IMS94 — APPROVED** [review]

**Pull request merged**